llvm/docs/GlobalISel/MIRPatterns.rst

```diff
@@ -143,6 +143,57 @@ Semantics:
     (apply (G_FSUB $dst, $src, $tmp),
            (G_FNEG GITypeOf<"$dst">:$tmp, $src))>;
 
+GIVariadic
+~~~~~~~~~~
+
+``GIVariadic<>`` is a ``GISpecialType`` that allows for matching 1 or
+more operands remaining on an instruction.
+
+Type Parameters:
+
+* The minimum number of additional operands to match. Must be greater than zero.
+
+  * Default is 1.
+
+* The maximum number of additional operands to match. Must be strictly greater
+  than the minimum.
+
+  * 0 can be used to indicate there is no upper limit.
+  * Default is 0.
+
+Semantics:
+
+* ``GIVariadic<>`` operands can not be defs.
+* ``GIVariadic<>`` operands can only appear as the last operand in a 'match' pattern.
+* Each instance within a 'match' pattern must be uniquely named.
+* Re-using a ``GIVariadic<>`` operand in an 'apply' pattern will result in all
+  the matched operands being copied from the original instruction.
+* The min/max operands will result in the matcher checking that the number of operands
+  falls within that range.
+* ``GIVariadic<>`` operands can be used in C++ code within a rule, which will
+  result in the operand name being expanded to an iterator range containing all
+  the matched operands, similar to the one returned by ``MachineInstr::operands()``.
+
+.. code-block:: text
+
+  def build_vector_to_unmerge: GICombineRule <
+    (defs root:$root),
+    (match (G_BUILD_VECTOR $root, GIVariadic<>:$args),
+           [{ return checkBuildVectorToUnmerge(${args}); }]),
```

Comment:
Include an example of unpacking the first N args? Like 
`G_BUILD_VECTOR $root, $srca, $srcb, ...:$remainder`

Reply:
That's not yet implemented for MIR patterns. In C++ you can just convert the iterator_range to an ArrayRef I think.
I need some ideas on how to proceed

What kind of operations do we want? 
 - Only head/tail of the list, or do we need specific indexes?
 - Do we need one element at a time, or multiple ?
 
Coming up with some example patterns would greatly help. I'm not sure how unpacking would be used in practice, because we don't have loops, so we can't do things like "for each element, emit this"

Reply:
Isn't that what's going in in VariadicTypeTestReuse?

Reply:
No, that just takes the N matched operands and puts them at a different position in the output instruction.
So `$b` can represent 2 to 4 operands, and the output instruction has $a, $b (so 2 to 4 operands being put here), and $c

It's a test to check that in apply patterns, a matched variadic can go anywhere, but in a match pattern it can only go at the end

Note that this whole thing is very much a prototype implementation, I put it up for review to discuss things like this and refine it

Reply:
Could you spell out the signature of `checkBuildVectorToUnmerge`? In particular, I am interested in the argument type.

Reply:
I thought `ArrayRef<MachineOperand>` would work out of the box but to my surprise, `ArrayRef` did not have a constructor from `iterator_range`. I tried adding one, but after 10 minutes of fighting templates (because we'd need one for both const and non-const ranges, and it apparently doesn't work), I gave up and made a helper function to do the conversion.

So now it's really just a `ArrayRef<MachineOperand>`, created through `GIMatchTableExecutor::getVariadicOperands`.